contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -2310,3 +2311,57 @@ sp_addlinkedsrvlogin_internal(PG_FUNCTION_ARGS)
 
 	return (Datum) 0;
 }
+
+Datum
+sp_dropserver_internal(PG_FUNCTION_ARGS)
+{
+	char *linked_srv = PG_ARGISNULL(0) ? NULL : text_to_cstring(PG_GETARG_VARCHAR_PP(0));
+	char *droplogins = PG_ARGISNULL(1) ? NULL : lowerstr(text_to_cstring(PG_GETARG_BPCHAR_PP(1)));
+
+	StringInfoData query;
+
+	if (linked_srv == NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_FDW_ERROR),
+					errmsg("@server parameter cannot be NULL")));
+	
+	initStringInfo(&query);
+
+	/*
+	 * We prepare the following query to drop foreign server. This will
+	 * be executed using ProcessUtility():
+	 *
+	 * DROP SERVER <servername> CASCADE
+	 *
+	 * linked logins along with server are dropped if @droplogins = 'NULL'
+	 * or @droplogins = 'droplogins' so we add CASCADE.
+	 */
+	if ((droplogins == NULL) || ((strlen(droplogins) == 10) && (strncmp(droplogins, "droplogins", 10) == 0)))
+	{
+		appendStringInfo(&query, "DROP SERVER \"%s\" CASCADE", linked_srv);
+
+		exec_utility_cmd_helper(query.data);
+		pfree(query.data);
+
+		if (linked_srv)
+			pfree(linked_srv);
+
+		if (droplogins)
+			pfree(droplogins);
+
+	}
+	else
+	{
+		if (linked_srv)
+			pfree(linked_srv);
+
+		if (droplogins)
+			pfree(droplogins);
+
+		ereport(ERROR,
+			(errcode(ERRCODE_FDW_ERROR),
+				errmsg("Invalid parameter value for @droplogins specified in procedure 'sys.sp_dropserver', acceptable values are 'droplogins' or NULL.")));
```

Comment:
initStringInfo would have allocated memory for query so it needs to be freed here before raising an error.

Reply:
We do not actually need to free since the context will be reset as part of this error. We do not free in such scenarios and there are multiple examples for this.